src/app/shared/wysiwyg/wysiwyg.component.ts

```diff
@@ -0,0 +1,399 @@
+        <button (click)="createLink()">Okay</button>
+        <button (click)="prompt.hide()">Cancel</button>
+      </div>
+    </publish-prompt>
+  `,
+  styleUrls: ['wysiwyg.component.css']
+})
+
+export class WysiwygComponent implements OnInit, OnChanges, OnDestroy {
+  @Input() model: BaseModel;
+  @Input() name: string;
+  @Input() images: ImageModel[];
+
+  @ViewChild('contentEditable') private el: ElementRef;
+  view: MenuBarEditorView;
+
+  @ViewChild('prompt') private prompt: PromptComponent;
+  @ViewChild('url') private url: ElementRef;
+  linkURL: string;
+  linkTitle: string;
+
+  ngOnInit() {
+    if (this.model) {
+      let state = EditorState.create(this.stateConfig());
+      this.view = new MenuBarEditorView(this.el.nativeElement, this.viewProps(state));
+    }
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    // once images are loaded and if images are added/removed from story, array length should change
+    let changed = changes['images'].currentValue.length !== changes['images'].previousValue.length;
+    if (this.view && changed) {
+      let state = this.view.editor.state.reconfigure(this.stateConfig());
+      this.view.update(this.viewProps(state));
+    }
+  }
+
+  ngOnDestroy() {
+    if (this.view) {
+      this.view.editor.destroy();
+    }
+  }
+
+  viewProps(state) {
+    return {
+      state,
+      onAction: (action) => {
+        this.view.updateState(this.view.editor.state.applyAction(action));
+        this.model.set(this.name, defaultMarkdownSerializer.serialize(this.view.editor.state.doc));
+      }
+    };
+  }
+
+  stateConfig() {
+    return {
+      doc: defaultMarkdownParser.parse(this.model[this.name] ? this.model[this.name] : ''),
+      plugins: this.buildMenuItemsPlugin()
+    };
+  }
+
+  get changed(): boolean {
+    return this.model.changed(this.name, false);
+  }
+
+  get invalid(): string {
+    return this.model.invalid(this.name);
+  }
+
+  cmdItem(cmd, options) {
+    let passedOptions = {
+      label: options.title,
+      run: cmd,
+      select(state) {
+        return cmd(state)
+      }
+    };
+    for (let prop in options) {
+      passedOptions[prop] = options[prop];
+    }
+    return new MenuItem(passedOptions);
+  }
+
+  markActive(state, type) {
+    let {from, to, empty} = state.selection;
+    if (empty) {
+      return type.isInSet(state.storedMarks || state.doc.marksAt(from));
+    } else {
+      return state.doc.rangeHasMark(from, to, type);
+    }
+  }
+
+  markItem(markType, options) {
+    let passedOptions = {
+      active: (state) => {
+        return this.markActive(state, markType);
+      }
+    };
+    for (let prop in options) {
+      passedOptions[prop] = options[prop];
+    }
+    return this.cmdItem(toggleMark(markType), passedOptions);
+  }
+
+  wrapListItem(nodeType, options) {
+    return this.cmdItem(wrapInList(nodeType, options.attrs), options);
+  }
+
+  isURLInvalid() {
+    return this.url.invalid && this.url.dirty;
```

Comment:
One thing I would probably try if I put in more work on the linking specifically is rather than just toggling off the mark/link when clicking link on already linked text, it would read the link and fill in the URL and title in the prompt. Seems like this would be more useful since there's no way for the user to verify a link after they have entered it, but it also seems like it could be painful to implement.